### PR TITLE
Cleanup structural solvers

### PR DESCRIPTION
**Pull request by philbucher**
This PR cleans the structural solvers and makes them consistent, e.g.
* solving #746  
* adding a linear analysis type to the dynamic solver (and removing the line search which makes no sense for a dynamic analysis)
* adding the eigensolver to the python wrapper  
* proper error throwing with listing of available options
* tests have been updated
* refactoring to avoid code duplication
* The user is forced to set `parallel_type`, `solver_type`, `analysis_type` (and for dynamic analysis `time_integration_method`) This prevents the silent construction of solvers based on default parameters, which can be dangerous!
* ...

The biggest change is that the `solver_type` is changed for a dynamic analysis from `Implicit` to `Dynamic`. This is the API Breaker
The time integration (Implicit or Explicit) is chosen based on the `time_integration_method`. The `solution_type` is removed completely, bcs for a static simulation both `solver_type` and `solution_type` had to be set to `Static` and this setting was also not really used.

Sorry this is such a big change but it is really necessary! We discussed it at the chair and decided to do it bcs the current status is just confusing!

@jginternational the problemtype needs to be updated for this to work. I will make a separate issue for this

@KratosMultiphysics/structural-mechanics

### CODE REVIEW

**Review by loumalouomega — APPROVED** [review]
OK, wait for more reviews

**Review by KlausBSautter — APPROVED** [review]

**Comment by msandre** [comment]
It looks reasonable. It will probably result in a merge conflict with #813 where the eigen solver generation is moved to a factory. Whatever pr goes through first, the other will probably need to resolve this.

**Comment by msandre** [comment]
By the way, it looks like you changed a lot of functions for creating linear solver , etc. What was the reason for this?

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py
Comment (adityaghantasala) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py]: "multi_point_constraints_used"  is a flag. 
Also "ResidualBasedBlockBuilderAndSolverWithMpc" is also a block builder and solver.
I think the following should be the logic

```
if (block_builder){
   if (multi_point_constraints_used)
      make block builder and solver with MPC 
   else 
      make normal block builder and solver
} else {
     make elimination builder and solver 
}
```

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py
Comment (adityaghantasala) [inline applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: >  You must provide a "parallel_type" in your settings!
to 
> ERROR : "parallel_type" not specified in settings.
??
similarly for the other Exceptions in the following lines of code ??

**Review by adityaghantasala — CHANGES_REQUESTED** [review]
Since MPC builder and solver is also a block builder and solver, check for the usage of MPC constraints is to be done inside the condition for block_builder.

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py]: Fixed

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: Yes, I check for it since it is needed for the solver selection

**Comment by philbucher** [comment]
@msandre I see the conflict with #813, but I think the changes are small (fyi @loumalouomega )

I haven't touched the linear solver creation, I changed how the mechanical solver is created. This way it was e.g. it was not noticed that the dynamic solver didn't have a linear solver.
Before each solver could do it's own thing, now the basic version is providing the linear (linear_strategy) and nonlinear (newton_raphson_strategy) solver. If desired, the mechanical solver creation can still be overridden, as done by e.g. the static solver, to provide more functionalities (in case of the static solver: FormFinding, Arc-Length and line search)

Furthermore, the solvers are divided into static/dynamic, therefore both get the analysis_type (linear or non-linear) through the base class. The difference is the scheme, thats why I made this the "abstract" class that each solver has to implement

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_eigensolver.py
Comment (msandre) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_eigensolver.py]: I think it is not necessary to specify EigenValue in the scheme if we already know it's eigenvalue from the solver_type.

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py
Comment (msandre) [inline applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: Because this pr will break existing cases by changing the json format, I would like to push that we also clean up some inconsistencies with CamelCase versus lower_case_with_underscores. From the wiki page style guide: "The name of the property should be written in lowercase with underscores "_" in between words if necessary". Unless I am not understanding something about the convention, can you make Dynamic, Static, EigenValue, etc. lower case with underscores?

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py
Comment (msandre) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:44]: I will not insist on this, but I prefer using the python script name in "solver_type" compared to custom names like Static, Dynamics, etc. My argument for this is that if you use the script name:

1) You don't need to search in the code for which python solver file gets loaded. It tells you directly without some intermediate keyword.

2) The MainKratos.py doesn't need to be modified since you tell it directly what solver to load. Everything related to the analysis type is left to that solver. With keywords like Static, the MainKratos.py already needs to know that this is a structural analysis and check for the solver to load.

**Review by msandre — CHANGES_REQUESTED** [review]

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: fixed

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_eigensolver.py
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_eigensolver.py]: fixed

**Comment by loumalouomega** [comment]
I have been speaking with @rubenzorrilla and in order to keep using the wrapper, and then being able to use MPI directly from the Parameters.json we should keep this compatibility. It is still possible to use MPI just changing one line in the Parameters.json?

**Comment by philbucher** [comment]
@loumalouomega and @rubenzorrilla I have not changed the behavior of the parallelism. It is (still) the same as in the FluidApp. So yes, you can still select the parallelism by changing one line in json

**Review by msandre — APPROVED** [review]

**Comment by philbucher** [comment]
@RiccardoRossi do you have any objections / comments? Otherwise I would merge beginning of next week

File: applications/StructuralMechanicsApplication/python_scripts/convergence_criteria_factory.py
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/python_scripts/convergence_criteria_factory.py]: I would throw an error if none of the options is recognized

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: Isn't this verified by the "Validate" function?

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: same

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: same

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: same

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_eigensolver.py
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_eigensolver.py]: please use the linar_solver_factory here to construct the linear solver settings

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_implicit_dynamic_solver.py
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_implicit_dynamic_solver.py:112]: i would say that line search is a valid option, at least for the static case. Isthat not so?

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_static_solver.py
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_static_solver.py:94]: mmm

all of this of the update_arc_lenght should not be done in python ... it should be a function in c++.
Maybe a function can be created in VariableUtils for the purpose

File: applications/StructuralMechanicsApplication/tests/patch_test/updated_lagrangian/patch_test_3D_shear_hexa_parameters.json
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/tests/patch_test/updated_lagrangian/patch_test_3D_shear_hexa_parameters.json:10]: I just wonder if we need all of those files. We have one simple python_only file that includes everything, without json and parameters. Isn't that better instead of having a lot of mdpas and parameters to only have that?

to make it clear this is a question. Opinions are more than welcome

**Review by RiccardoRossi — APPROVED** [review]
some minor comments. Otherwise i approve.

HOWEVER before merging ensure that the corresponding changes are done in the python interface

**Review by adityaghantasala — APPROVED** [review]

File: applications/StructuralMechanicsApplication/python_scripts/convergence_criteria_factory.py
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/convergence_criteria_factory.py]: Done

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: You are right, I fixed it

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: fixed

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: fixed

File: applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py]: fixed

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_eigensolver.py
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_eigensolver.py]: done

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_implicit_dynamic_solver.py
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_implicit_dynamic_solver.py:112]: In the static case it can be used, it is still available in this case.
Only in a dynamic Analysis it doesn't make sense

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_static_solver.py
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_static_solver.py:94]: Will do in separate PR

File: applications/StructuralMechanicsApplication/tests/patch_test/updated_lagrangian/patch_test_3D_shear_hexa_parameters.json
Comment (philbucher) [reply applications/StructuralMechanicsApplication/tests/patch_test/updated_lagrangian/patch_test_3D_shear_hexa_parameters.json:10]: Thats just the old format of the tests, isn't it?

**Comment by philbucher** [comment]
I plan to merge this on Tuesday in case someone still has suggestions for improvement
After that we can focus on [updating the GiD Interface](https://github.com/KratosMultiphysics/GiDInterface/issues/159) to be compliant with these changes for the Release!

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_static_solver.py
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_static_solver.py:94]: @RiccardoRossi There is only one test that uses the arc-length method, but it is outdated. I tried fixing it but didn't succeed. It seems like there is some more work necessary
Therefore I suggest doing it all together at some point, both fixing the test and implementing the function in c

**Comment by pooyan-dadvand** [comment]
@philbucher and @AndreasWinterstein would you please all issues and pull requests related to the features you want to include in release in the [milestone](https://github.com/KratosMultiphysics/Kratos/milestone/4)?

**Pull request merged**
